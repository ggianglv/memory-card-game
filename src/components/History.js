import React from "react";
import { LEVEL } from "../constants";
import { AppContext } from "../context";
import { formatTime } from "../utils/data";

const History = () => {
  const { state } = React.useContext(AppContext);
  const { history = [] } = state;

  const modeMap = {
    [LEVEL.EASY]: "Easy",
    [LEVEL.MEDIUM]: "Med",
    [LEVEL.HARD]: "Hard",
  };
  return (
    <div className="play-history">
      {history.map((item, index) => (
        <div key={index} className="modal-item history">
          <div className="history-item">
            <div className="label">Move:</div>
            <div className="value">{item.move}</div>
          </div>

          <div className="history-item">
            <div className="label">Time:</div>
            <div className="value">{formatTime(item.time)}</div>
          </div>

          <div className="history-item">
            <div className="label">Mode:</div>
            <div className="value">{modeMap[item.mode]}</div>
          </div>
        </div>
      ))}
    </div>
  );
};

export default History;
