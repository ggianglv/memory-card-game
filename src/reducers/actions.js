export const SET_MODE = "SET_MODE";
export const UPDATE_TIME = "UPDATE_TIME";
export const UPDATE_MOVE = "UPDATE_MOVE";
export const UPDATE_DATA = "SET_INIT_DATA";
export const UPDATE_START = "UPDATE_START";
export const UPDATE_TOPIC = "UPDATE_TOPIC";
export const UPDATE_REMOVED = "UPDATE_REMOVED";
