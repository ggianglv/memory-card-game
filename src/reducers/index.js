import { LEVEL, TOPICS } from "../constants";
import {
  UPDATE_DATA,
  SET_MODE,
  UPDATE_MOVE,
  UPDATE_REMOVED,
  UPDATE_START,
  UPDATE_TIME,
  UPDATE_TOPIC,
} from "./actions";
import { generateGameData } from "../utils/data";

const initData = generateGameData({ topic: TOPICS.RICK, mode: LEVEL.EASY });

export const initialState = {
  move: 0,
  removed: {},
  time: 0,
  start: false,
  topic: TOPICS.RICK,
  mode: LEVEL.EASY,
  isLoading: false,
  complete: false,
  history: [],
  ...initData,
};

export const appReducer = (state, action) => {
  switch (action.type) {
    case SET_MODE:
      const data = generateGameData({
        topic: state.topic,
        mode: action.payload,
      });
      return {
        ...state,
        ...data,
        mode: action.payload,
      };
    case UPDATE_DATA:
      return {
        ...state,
        ...action.payload,
      };
    case UPDATE_START:
      return {
        ...state,
        start: action.payload,
      };

    case UPDATE_MOVE:
      return {
        ...state,
        move: action.payload,
      };
    case UPDATE_TIME:
      return {
        ...state,
        time: state.time + 1,
      };
    case UPDATE_REMOVED:
      return {
        ...state,
        removed: action.payload,
      };
    case UPDATE_TOPIC: {
      const data = generateGameData({
        topic: action.payload,
        mode: state.mode,
      });
      return {
        ...state,
        ...data,
        topic: action.payload,
        start: false,
        time: 0,
        move: 0,
      };
    }
    default:
      return state;
  }
};
