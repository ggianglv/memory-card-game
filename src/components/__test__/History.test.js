import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import { AppContext } from "../../context";
import { initialState } from "../../reducers";
import History from "../History";
import renderer from "react-test-renderer";

test("render History without crash", () => {
  render(
    <AppContext.Provider value={{ state: initialState, dispatch: () => {} }}>
      <History />
    </AppContext.Provider>
  );
});

test("History to match snapshot", () => {
  const tree = renderer
    .create(
      <AppContext.Provider value={{ state: initialState, dispatch: () => {} }}>
        <History />
      </AppContext.Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
