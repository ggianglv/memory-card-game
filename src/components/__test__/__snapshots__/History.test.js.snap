// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History to match snapshot 1`] = `
<div
  className="play-history"
/>
`;
